File: 60058.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

string solveRecursively(string p) {
    string u = "";
    string v = "";
    
    int open = -1;
    int close = -1;
    
    if(p == "") {
        return p;
    }
    
    for(int i = 0; i < p.size(); i++) {
        if(p[i] == '(') {
            open++;
        } else if(p[i] == ')') {
            close++;
        }
        
        if(open == close && open != -1) {
            u = string(p.begin(), p.begin() + i + 1);
            v = string(p.begin() + i + 1, p.end());
            break;
        }
    }
    
    // u가 올바른경우 v에 대해 재귀
    vector<char> temp;
    bool isCorrect = false;
    for(auto each : u) {
        if(each == '(') {
            temp.push_back(each);
        } else if(each == ')') {
            if(temp.empty()) {
                isCorrect = false;
                break;
            } else {
                temp.pop_back();
                isCorrect = true;
            }
        }
    }
    if(isCorrect) {
        return u + solveRecursively(v);
    }
    
    string trimmedU = string(u.begin() + 1, u.end() - 1);
    string reversedU = "";
    
    for(auto each : trimmedU) {
        if(each == '(') {
            reversedU.push_back(')');
        } else if(each == ')') {
            reversedU.push_back('(');
        }
    }
    
    return "(" + solveRecursively(v) + ")" + reversedU;
}

string solution(string p) {
    string answer = "";

    answer = solveRecursively(p);
    
    return answer;
}

File: 60058.py
def solution(p):
    answer = solveRecursively(p)
    return answer


def solveRecursively(p):
    u = ''
    v = ''
    
    if len(p) == 0 :
        return ""
    
    for index in range(len(p)):
        temp = p[0:index + 1]
        if temp.count('(') == temp.count(')'):
            u = temp
            v = p[index+1:]
            break
            
    stack = []
    isCorrect = True
    for each in u:
        if each == '(':
            stack.append('(')
        elif each == ')':
            if len(stack) == 0:
                isCorrect = False
                break
            else:
                stack.pop()
    
    if isCorrect:
        return u + solveRecursively(v)
    
    trimmedU = u[1:-1]
    modifiedU = ''
    for each in trimmedU:
        if each == ')':
            modifiedU += '('
        elif each == '(':
            modifiedU += ')'
            
    return '(' + solveRecursively(v) + ')' + modifiedU
